# Bitwise **NOT** Operator: `~x`
---

- `~x` = 1's complement of ***x***


1. Find Complement of ***x***: 
    - flip all bits (includin sign)
    - get intermediate result
2. Flip inetermediate result and add/subtract 1 (depending on sign)
3. Result = Complement 


- Python Does Not Use A *Sign Bit*
    - `-13 = -0b1101`
    - stored externally 

### Positive Start 
$\begin{array}{c|c|llll|c}
{step}&{sign}&8&4&3&1&decimal\\
\hline
\hline
{original}&+&1&1&0&1&+13\\
{flip}\\
\hline
{comp1}&-&0&0&1&0&-2\\
{flip}&+&1&1&0&1&+13\\
{add 1}&&&&+&1\\
\hline
{comp2}&-&1&1&1&0&-14\\
\end{array}$

### Negative Start 
$\begin{array}{c|c|llll|c}
{step}&{sign}&8&4&3&1&decimal\\
\hline
\hline
{original}&-&1&1&1&0&-14\\
{flip}\\
\hline
{comp1}&+&0&0&0&1&+1\\
{flip}&-&1&1&1&0&-14\\
{minus 1}&&&&-&1\\
\hline
{comp2}&+&1&1&0&1&+13\\
\end{array}$

In [70]:
pos_13, neg_13 = 13, -13
pos_14, neg_14 = 14, -14

In [71]:
#FLIPPING
def flip(c): 
    if c == '1': return '0'
    elif c == '0': return '1'
    else: return "error"
def sign_print(sign): 
    if sign == '0': return "+"
    else: return "-"

In [83]:
# 1's COMPLEMENT OF A BINARY NUMBER
def comp1(decimal): 
    if decimal >= 0: 
        s = '0'
        b = bin(decimal)[2:]
    else: 
        s = '1' 
        b = bin(decimal)[3:]
    print(f"\nOriginal Decimal: {decimal} \nOriginal Binary:  {b} \n Sign: {s}, {sign_print(s)}")

    ones = ""
    for i in range(len(b)): 
        ones += flip(b[i])
    s = flip(s)
    print(f"Complement 1: {ones}, {int(ones,2)} \n Sign: {s}, {sign_print(s)}")

    return s, ones

p13_1_s, p13_1_ones = comp1(pos_13)



Original Decimal: 13 
Original Binary:  1101 
 Sign: 0, +
Complement 1: 0010, 2 
 Sign: 1, -


In [86]:
# 1's COMPLEMENT FLIPPED - to find 2's COMPLEMENT 
def comp1_flip(s, c1):
    twos = list(c1)
    for i in range(len(twos)): 
        twos[i] = flip(twos[i])
    s = flip(s)
    print(f"Complement 1 FLIP PREP: {''.join(twos)} \n Sign: {s}, {sign_print(s)}")
    
    return s, twos 

p13_1flip_s, p13_1flip_twos = comp1_flip(p13_1_s, p13_1_ones)
    

Complement 1 FLIP PREP: 1101 
 Sign: 0, +


In [136]:
# ADD ONE TO 1'S COMPLEMENT 
def comp2_add(s,twos):
    addy = '1'.zfill(len(twos))
    result = ''
    carry = 0 
    if s == '0': 
        for i in range(len(twos)-1, -1,-1): 
            temp = carry 
            temp += 1 if twos[i] == '1' else 0 
            temp += 1 if addy[i] == '1' else 0 

            result = ('1' if temp % 2 == 1 else '0') + result 
            carry = 0 if temp < 2 else 1 
        if carry != 0: 
            result = '1' + result 
        s = flip(s)
        print(f"ADD 1 -> {result}")
        print(f" Sign: {s}, {sign_print(s)}")
        print(f" {sign_print(s)}{'0b'+result} -> {sign_print(s)}{int(result,2)}")

    return s, twos

p13_comp2_s, p13_comp2_twos = comp2_add('0', '1101')
    

ADD 1 -> 1110
 Sign: 1, -
 -0b1110 -> -14


In [121]:
# SUBTRACT ONE FROM 1'S COMPLEMENT 


# SOMETHING IS NOT WORKING 

def comp2_sub(s,twos): 
    result = '0'
    for i in range(len(twos)-1, -1,-1): 
        if twos[i] == '1': 
              result = '0' + result 
              result = twos[:i-1]
              #result = twos[:i-1] + '0' + result
              break 
        else: 
            if twos[i-1] == '1': 
                 result = '01' + result
                 result = twos[i-2] + result
                 #twos[i] = '1' 
                 #twos[i-1] = '0' 
                 break 
            else: 
                 i -= 1 
        #result = ''.join(twos)
    s = flip(s)
    print(f"SUBTRACT 1 -> {result}")
    print(f"Sign: {s}, {sign_print(s)} -> 0b{result}")
    return s,result 

p13_final_s, p13_final = comp2_sub('1','1100')

SUBTRACT 1 -> 1010
Sign: 0, + -> 0b1010


In [141]:
#d = 13
d = -10
step_1s, step_1bin = comp1(d)
step_2s, step_2bin = comp1_flip(step_1s,step_1bin)
#step_3s_add, step_3bin_add = comp2_add(step_2s, step_2bin)

print("\n PICKUP HERE - NOT WORKING RIGHT")
step_3s_sub, step_3bin_sub = comp2_sub(step_2s, step_2bin)
print(f"Bitwise Operator ~{d}: {~d}")



Original Decimal: -10 
Original Binary:  1010 
 Sign: 1, -
Complement 1: 0101, 5 
 Sign: 0, +
Complement 1 FLIP PREP: 1010 
 Sign: 1, -

 PICKUP HERE - NOT WORKING RIGHT
SUBTRACT 1 -> 0010
Sign: 0, + -> 0b0010
Bitwise Operator ~-10: 9


### BIG BOY

In [112]:
def complement_1(decimal):
    if decimal >= 0: 
        s = '0'
        b = bin(decimal)[2:]
    else: 
        s = '1' 
        b = bin(decimal)[3:]

    def flip(c): 
        if c == '1': return '0'
        elif c == '0': return '1'
        else: return "error"
    def sign_print(sign): 
        if sign == '0': return "+"
        else: return "-"

    ones,twos = "",""
    print(f"\nOriginal Decimal: {decimal} \nOriginal Binary: {b} \n Sign: {s}, {sign_print(s)}")


    # 1's Complement 
    for i in range(len(b)): 
        ones += flip(b[i])
    s = flip(s)
    print(f"Complement 1: {ones} \n Sign: {s}, {sign_print(s)}")

    # 2's Complement
    twos = list(ones)
    for i in range(len(twos)):
        twos[i] = flip(twos[i])
    s = flip(s)
    print(f"Complement 2 FLIP PREP: {''.join(twos)} \n Sign: {s}, {sign_print(s)}")

    # Add 1
    addy = '1'.zfill(len(b))
    result = ''
    carry = 0 
    if s == '0':
        for i in range(len(twos) - 1, -1,-1): 
            temp = carry 
            temp += 1 if twos[i] == '1' else 0 
            temp += 1 if addy[i] == '1' else 0 
            result = ('1' if temp % 2 == 1 else '0') + result 

            carry = 0 if temp < 2 else 1 
        if carry != 0: 
            result = '1' + result 
        print(f"ADD 1 -> {result}")
        s = flip(s)
        print(f" Sign: {s}, {sign_print(s)} -> -0b{result}")
        print(f"~{decimal} -> {bin(decimal)} = {'0b'+result} -> {sign_print(s)}{int(result,2)}")

    
    #PICKUP HERE 
    else: 
        #sub_check = []
        for i in range(len(twos)-1, -1,-1): 
            if twos[i] == '1': 
                twos[i] = '0'
                #sub_check.append(twos[i])
                break 
            else: 
                if twos[i-1] == '1': 
                    twos[i] = '1' 
                    twos[i-1] = '0' 
                    #sub_check.append(twos[i])
                    break
                else: 
                    #sub_check.append('-1')
                    i -= 1 
        result = ''.join(twos)
        #print(sub_check)
        print(f"SUBTRACT 1 -> {result}")
        s = flip(s)
        print(f"Sign: {s}, {sign_print(s)} -> 0b{result}")
        print(f"~{decimal} -> {bin(decimal)} = {'0b'+result} -> {sign_print(s)}{int(result,2)}")

    return ''.join(twos)


pos1 = complement_1(13)
neg1 = complement_1(-14)
print(pos1, neg1)


Original Decimal: 13 
Original Binary: 1101 
 Sign: 0, +
Complement 1: 0010 
 Sign: 1, -
Complement 2 FLIP PREP: 1101 
 Sign: 0, +
ADD 1 -> 1110
 Sign: 1, - -> -0b1110
~13 -> 0b1101 = 0b1110 -> -14

Original Decimal: -14 
Original Binary: 1110 
 Sign: 1, -
Complement 1: 0001 
 Sign: 0, +
Complement 2 FLIP PREP: 1110 
 Sign: 1, -
SUBTRACT 1 -> 1101
Sign: 0, + -> 0b1101
~-14 -> -0b1110 = 0b1101 -> +13
1101 1101


In [3]:
a,b = 5,21
na, nb = bin(a)[2:], bin(b)[2:]

print("binary a: {}, {}\n OG:  {} \nNOT: {} \nSUM: {} \n".format(bin(a),na,a,(~a), a+~a))
print("binary b: {}, {}\n OG:  {} \nNOT: {} \nSUM: {}".format(bin(b),nb,b,(~b), b+~b))


binary a: 0b101, 101
 OG:  5 
NOT: -6 
SUM: -1 

binary b: 0b10101, 10101
 OG:  21 
NOT: -22 
SUM: -1


In [4]:
def check_bitwise_not(list): 
    for i in range(len(list)):
        print("binary {}: {}".format(list[i],bin(list[i])))
        print(" NOT: {}".format(~list[i]))

not_list = [15,-15]
check_bitwise_not(not_list)

binary 15: 0b1111
 NOT: -16
binary -15: -0b1111
 NOT: 14


### Numbers Stored in Binary Format 
 - __ ________ Leftmost Bit Determines Sign
    - 0 -> Positive
    - 1 -> Negative
- x = 13 -> binary_x = **0** 1101
    - leftmost reserved for sign, `0 = positive`
> #### Original: `~13` or  `0 | 1101`
> #### 1's Complement: `1 | 0010` flip everything including sign 
> #### 2's Complement: `1 | 1101` flip 1's complement - except sign 
>> #### ADD ONE to 2's Complement: ($1+1=10$)
>>>
>>> $1 | 1110= \begin{vmatrix} -&0&0&1&0 \cr +&1&1&0&1 \cr &&&+&1 \end{vmatrix}$

### Positive Start 
$\begin{array}{c|c|llll|c}
{step}&{sign}&8&4&3&1&decimal\\
\hline
\hline
{original}&+&1&1&0&1&+13\\
{flip}\\
\hline
{comp1}&-&0&0&1&0&-2\\
{flip}&+&1&1&0&1&+13\\
{add 1}&&&&+&1\\
\hline
{comp2}&-&1&1&1&0&-14\\
\end{array}$

### Negative Start 
$\begin{array}{c|c|llll|c}
{step}&{sign}&8&4&3&1&decimal\\
\hline
\hline
{original}&-&1&1&1&0&-14\\
{flip}\\
\hline
{comp1}&+&0&0&0&1&+1\\
{flip}&-&1&1&1&0&-14\\
{minus 1}&&&&-&1\\
\hline
{comp2}&+&1&1&0&1&+13\\
\end{array}$


### 1's Complement: Flipping Bits 
### 2's Complement: Add 1 to 1's Complement 
- negative number stored as two's complement 
- ADD 1: `1 + 1 = 10` -> `1101 + 1 = 1110`

In [5]:
not_zero, not_one = ~0, ~1
print(f"NOT Zero: ~0 = {not_zero} \nNOT One: ~1 = {not_one}")

NOT Zero: ~0 = -1 
NOT One: ~1 = -2


In [6]:
pos = 0b1101
neg = -0b1101
print(f"Positive: {pos} \nNegative: {neg}")

Positive: 13 
Negative: -13
